### ISSUE TITLE
Using impl class when configuring a provider

### ISSUE THREAD

**Issue by jerrinot** [hazelcast-guides/caching-springboot-jcache#4]
The guide says:
```
Finally, we need to specify the JCache provider in the application.properties file.
[...]
spring.cache.jcache.provider=com.hazelcast.cache.impl.HazelcastServerCachingProvider
```

The impl package inside a configuration makes me uneasy. Isnt there another way? If there isnt then please open a github issue against Hazelcast. Thank you!

**Comment by alparslanavci** [comment]
Starting with 4.x, when you call `Caching.getCachingProvider("com.hazelcast.cache.HazelcastCachingProvider")`, it returns the client provider by default since now the client classes are included within the Hazelcast package.

I guess that is why we need to specify `com.hazelcast.cache.impl.HazelcastServerCachingProvider` as the provider explicitly.

**Comment by pivovarit** [comment]
Shouldn't this be considered a bug then?

I was just trying passing `com.hazelcast.cache.HazelcastCachingProvider` as a provider, but it clearly doesn't work

**Comment by alparslanavci** [comment]
Yes, passing `com.hazelcast.cache.HazelcastCachingProvider` doesn't work. I talked to Vassilis about it, and I am not sure if it should be considered as a bug but we definitely need it to be mentioned in the 4.0 migration guide. 

BTW, it also affects spring-boot's Hazelcast 4.0 upgrade.

**Comment by jerrinot** [comment]
yes. it's a bug. under no circumstances we should be encouraging users to touch *.impl.* classes. 
If there is no other way then please open an issue against Hazelcast.

**Comment by pivovarit** [comment]
I'm actually reviewing the guide at the moment. Will figure out if there's some non-impl way of doing that.

**Comment by pivovarit** [comment]
Not only this seems to be the only way to make it work, but it's also quite commonly used:

https://github.com/search?q=spring.cache.jcache.provider%3Dcom.hazelcast.cache.impl.HazelcastServerCachingProvider+extension%3Aproperties&type=Code&ref=advsearch&l=&l=

**Comment by jerrinot** [comment]
oh boy, we are teaching people bad practices :(

**Comment by pivovarit** [comment]
Opened this one: https://github.com/hazelcast/hazelcast/issues/17312

**Comment by pivovarit** [comment]
After the next Hazelcast release, we'll be able to replace usages of `impl.*` classes with public ones
